sw_i2c/sample-implementations/linux_user_space/sensirion_sw_i2c_implementation.c

```diff
@@ -0,0 +1,230 @@
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+
+#include <fcntl.h> /* close, open */
+#include <sys/types.h> /* mode_t */
+#include <stdio.h> /* fprintf, perror, stderr */
+#include <string.h> /* strlen */
+#include <unistd.h> /* access, lseek, read, usleep */
+#include <stdlib.h> /* exit */
+
+#include "sensirion_sw_i2c_gpio.h"
+#include "sensirion_arch_config.h"
+
+/*
+ * We use the following names for the two I2C signal lines:
+ * SCL for the clock line
+ * SDA for the data line
+ *
+ * Both lines must be equipped with pull-up resistors appropriate to the bus
+ * frequency.
+ */
+#define GPIO_PIN_SCL 12
+#define GPIO_PIN_SDA 13
+#define GPIO_DIR "/sys/class/gpio/"
+
+#define str(x) #x
+#define GPIO_ID "gpio"
+#define GPIO(p) GPIO_ID str(p)
+#define GPIO_PATH(p) GPIO_DIR p
+#define GPIO_EXPORT_PATH GPIO_PATH("export")
+#define GPIO_SCL_DIR GPIO_PATH(GPIO(GPIO_PIN_SCL))
+#define GPIO_SDA_DIR GPIO_PATH(GPIO(GPIO_PIN_SDA))
+#define GPIO_SCL_PATH(d) GPIO_SCL_DIR d
+#define GPIO_SDA_PATH(d) GPIO_SDA_DIR d
+#define GPIO_SCL_DIRECTION GPIO_SCL_PATH("/direction")
+#define GPIO_SDA_DIRECTION GPIO_SDA_PATH("/direction")
+#define GPIO_SCL_VALUE GPIO_SCL_PATH("/value")
+#define GPIO_SDA_VALUE GPIO_SDA_PATH("/value")
+
+#define GPIO_DIRECTION_IN "in"
+#define GPIO_DIRECTION_OUT "out"
+#define GPIO_LOW 0
+
+static int scl_dir_fd;
+static int scl_val_fd;
+static int sda_dir_fd;
+static int sda_val_fd;
+
+static int open_err(const char *path, mode_t mode)
+{
+    int fd = open(path, mode);
+    if (fd < 0) {
+        perror(NULL);
+        fprintf(stderr, "Error opening %s (mode %d)\n", path, mode);
+        exit(-1);
+    }
+    return fd;
+}
+
+static void rev_err(int fd)
+{
+    if (lseek(fd, 0, SEEK_SET) < 0) {
+        perror("Error seeking gpio");
+        exit(-1);
+    }
+}
+
+static void write_err(int fd, const char *buf)
+{
+    size_t len = strlen(buf);
+
+    if (write(fd, buf, len) != len) {
+        perror(NULL);
+        fprintf(stderr, "Error writing\n");
+        exit(-1);
+    }
+}
+
+static void gpio_export(const char *path, const char *export_pin)
+{
+    int fd;
+
+    if (access(path, F_OK) == -1) {
+        fd = open_err(GPIO_EXPORT_PATH, O_WRONLY);
+        write_err(fd, export_pin);
+        close(fd);
+    }
+}
+
+static void gpio_set_value(int fd, int value)
+{
+    char buf[] = {'0', '\0'};
+
+    buf[0] += value;
+    rev_err(fd);
+    write_err(fd, buf);
+}
+
+static void gpio_set_direction(int fd, const char *dir)
+{
+    rev_err(fd);
+    write_err(fd, dir);
+}
+
+static char gpio_get_value(int fd)
+{
+    char c;
+
+    rev_err(fd);
+    if (read(fd, &c, 1) != 1) {
+        perror("Error reading GPIO value");
+        exit(-1);
+    }
```

Comment:
Maybe create a `read_xxx` function as well?

Reply:
Also thought about it, but it's only used here and I don't see this changing